doc/developer-guides/sw_design_guidelines.rst

```diff
@@ -0,0 +1,494 @@
+   |                    |           | different board configs,   | VM configurations are.    |                         |
+   |                    |           | they are defined           | Thus hypervisor does not  |                         |
+   |                    |           | as static structure.       | need any additional       |                         |
+   |                    |           |                            | mechanism.                |                         |
+   +--------------------+-----------+----------------------------+---------------------------+-------------------------+
+   | Configuration data | N/A       | No.                        | The hypervisor shall use  | The maximum number of   |
+   | of the hypervisor  |           | The related pre-conditions | the internal resource/data| PCI devices in the VM,  |
+   |                    |           | are required.              | directly.                 | defined with            |
+   |                    |           | The design will guarantee  |                           | CONFIG_MAX_PCI_DEV_NUM  |
+   |                    |           | the correctness and this   |                           | through configuration.  |
+   |                    |           | shall be verified manually.|                           |                         |
+   +--------------------+-----------+----------------------------+---------------------------+-------------------------+
+
+
+Examples
+--------
+
+Here are some examples to illustrate when error handling codes are required on
+module level.
+
+**Example_1: Analyze the function 'partition_mode_vpci_init'**
+
+.. code-block:: c
+
+  /**
+   * @pre vm != NULL
+   * @pre vm->vpci->pci_vdev_cnt <= CONFIG_MAX_PCI_DEV_NUM
+   */
+  static int32_t partition_mode_vpci_init(const struct acrn_vm *vm)
+  {
+          struct acrn_vpci *vpci = (struct acrn_vpci *)&(vm->vpci);
+          struct pci_vdev *vdev;
+          struct acrn_vm_config *vm_config = get_vm_config(vm->vm_id);
+          struct acrn_vm_pci_ptdev_config *ptdev_config;
+          uint32_t i;
+
+          vpci->pci_vdev_cnt = vm_config->pci_ptdev_num;
+
+          for (i = 0U; i < vpci->pci_vdev_cnt; i++) {
+                  vdev = &vpci->pci_vdevs[i];
+                  vdev->vpci = vpci;
+                  ptdev_config = &vm_config->pci_ptdevs[i];
+                  vdev->vbdf.value = ptdev_config->vbdf.value;
+
+                  if (vdev->vbdf.value != 0U) {
+                          partition_mode_pdev_init(vdev, ptdev_config->pbdf);
+                          vdev->ops = &pci_ops_vdev_pt;
+                  } else {
+                          vdev->ops = &pci_ops_vdev_hostbridge;
+                  }
+
+                  if (vdev->ops->init != NULL) {
+                          if (vdev->ops->init(vdev) != 0) {
+                                  pr_err("%s() failed at PCI device (vbdf %x)!",
```

Comment:
If this error is detected, should the loop continue?  Should the function still exit with a ``return 0``?

Reply:
Based on current design, the loop will continue and the next vdev will be initialized. Failure on one vdev does not block other devices.